File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/headers/HeadersSpec.java
package io.micronaut.nats.docs.headers;

import io.micronaut.context.annotation.Property;
import io.micronaut.nats.testcontainers.Nats;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import io.nats.client.impl.Headers;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Arrays;
import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@MicronautTest
@Property(name = "spec.name", value = "HeadersSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class HeadersSpec implements TestPropertyProvider {

    @Test
    void testPublishingAndReceivingHeaders(ProductClient productClient, ProductListener productListener) {
// tag::producer[]
        productClient.send("body".getBytes());
        productClient.send("medium", 20L, "body2".getBytes());
        productClient.send(null, 30L, "body3".getBytes());

        Headers headers = new Headers();
        headers.put("productSize", "large");
        headers.put("x-product-count", "40");
        productClient.send("body4".getBytes(), headers);
        productClient.send("body5".getBytes(), Arrays.asList("xtra-small", "xtra-large"));
// end::producer[]

        await().atMost(60, SECONDS).until(() ->
                productListener.messageProperties.size() == 6 &&
                productListener.messageProperties.contains("true|10|small") &&
                productListener.messageProperties.contains("true|20|medium") &&
                productListener.messageProperties.contains("true|30|null") &&
                productListener.messageProperties.contains("true|40|large") &&
                productListener.messageProperties.contains("true|20|xtra-small") &&
                productListener.messageProperties.contains("true|20|xtra-large")
        );
    }

    @Override
    public Map<String, String> getProperties() {
        return Nats.getProperties();
    }
}

File: docs-examples/example-python/src/test/python/micronaut/nats/docs/headers/HeadersSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test
try:
    from io.nats.client.impl import Headers
except ImportError:  # TODO(python): packages under `io.` other than `io.micronaut` cannot be imported at runtime
    from nats.client.impl import Headers

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="HeadersSpec")
@MicronautTest(environments=["nats"])
class HeadersSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_publishing_and_receiving_headers(self):
        # tag::producer[]
        self.product_client.send("body".encode())
        self.product_client.send_with_headers("medium", 20, "body2".encode())
        self.product_client.send_with_headers(None, 30, "body3".encode())

        headers = Headers()
        headers.put("productSize", "large")
        headers.put("x-product-count", "40")
        self.product_client.send_with_nats_headers("body4".encode(), headers)
        self.product_client.send_with_product_sizes("body5".encode(), ["xtra-small", "xtra-large"])
        # end::producer[]

        expected = {"true|10|small", "true|20|medium", "true|30|null", "true|40|large", "true|20|xtra-small", "true|20|xtra-large"}
        for _ in range(600):
            if self.product_listener.message_properties == expected:
                break
            sleep(0.1)
        assert self.product_listener.message_properties == expected
